interp/src/primitives/stateful.rs

```diff
@@ -16,16 +18,21 @@ where
     })
 }
 
-//pipelined multiplication, but as of now takes one cycle
+///Pipelined Multiplication (3 cycles)
+///Still bounded by u64.
+///How to use:
+///[execute] with the desired bindings. To capture these bindings
+///into the internal (out) queue, [do_tick()].
+///The product associated with a given input will
+///be output on the third [do_tick()].
+///Note: Calling [execute] multiple times before [do_tick()] has no effect; only
+///the last set of inputs prior to the [do_tick()] will be saved.
 #[derive(Default)]
 pub struct StdMultPipe {
     pub width: u64,
     pub product: Value,
     update: Option<Value>,
-    // right now, trying to be 1 cycle, so no need for these
-    // left: Value,
-    // right: Value,
-    //cycle_count: u64, //0, 1, 2
+    queue: VecDeque<Option<Value>>, //invariant: always length 3.
```

Comment:
This could be a three element array instead, maybe a wrapper around such that implements the buffer shifts. Not sure that the full generality of a VecDeque is needed and it may be easier for the invariant to be impossible to violate.